src/graph.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding utf-8 -*-
+
+
+class Graph(object):
+
+    def __init__(self, initial_graph=None):
+        '''Pass in a graph in the form of a dictionary with the nodes as
+        key, and a list of edges as value, for example
+        {1:[2, 3]},
+         2:[]},
+         3:[1]}
+        '''
+        if initial_graph is None:
+            self.graph = {}
+        else:
+            self.graph = initial_graph
+
+    def __iter__(self):
+        return iter(self.graph)
+
+    def nodes(self):
+        '''return a list of all nodes in the graph'''
+        return self.graph.keys()
+
+    def edges(self):
+        '''returns a list of tuples, that are edges'''
+        return self.graph.values()
+
+    def add_node(self, n):
+        '''takes node and adds it to the graph'''
+        if n in self.graph:
+            raise ValueError('Node already exists please try again')
+        else:
+            self.graph[n] = []
```

Comment:
This works, but there's actually a better way than using an `if...else...` block. Dictionaries have a method called `setdefault()` that will set a value on a given key if that key is not in the dictionary, and if it is in the dictionary it'll simply leave it alone.